### PR TITLE
Implement unified way to generate `vcs_url` for the main package purl

### PR DESCRIPTION
**Pull request by chmeliik**
In preparation for the npm (and also pip) purl implementation

This should make it easier to generate purls for the main packages (which should use `vcs_url` qualifiers)

Also use this mechanism for gomod (when getting the repo name for the source dir)

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [x] New code has type annotations
- [n/a] Docs updated (if applicable)
- [n/a] Docs links in the code are still valid (if docs were updated)

### CODE REVIEW

**Comment by chmeliik** [comment]
@ejegrova FYI this won't be needed for pip git dependencies, since the pip git urls are already in the correct format for `vcs_url`. Or rather, the purl spec defines the format as "as defined in Python pip or the SPDX specification" :smile: 

But it should be useful for the main package

File: cachi2/core/scm.py [hunk 1]
```diff
@@ -1,16 +1,81 @@
...
+    url = _canonicalize_origin_url(origin.url)
+    commit_id = repo.head.commit.hexsha
+    return RepoID(url, commit_id)
+
+
+def _canonicalize_origin_url(url: str) -> str:
+    if "://" in url:
+        return url
+    elif re.match("^[^/]*:", url):
+        # scp-style is "only recognized if there are no slashes before the first colon"
```
Comment (brunoapimentel) [inline cachi2/core/scm.py:67]: Not sure if it's just me, but this comment would probably be more helpful if it was on top of the `elif`. 
(I tested the regex to figure out what it did, only to realize this comment explained it later :face_exhaling: )

File: cachi2/core/scm.py [hunk 2]
```diff
@@ -1,16 +1,81 @@
...
+
+    def as_vcs_url_qualifier(self) -> str:
+        """Turn this RepoID into a 'vcs_url' qualifier as defined by the purl spec.
+
+        See https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst#known-qualifiers-keyvalue-pairs
+        """
+        return f"git+{self.origin_url}@{self.commit_id}"
+
+
+def get_repo_id(repo: Union[str, PathLike[str], Repo]) -> RepoID:
```
[the remark this answers is not in the document]
Comment (chmeliik) [reply cachi2/core/scm.py:37]: If we assume that a path will always have the same repo id, we could

IMO it's not this function's place to assume that, but in practice it's unlikely to cause any issues

**Review by taylormadore — APPROVED** [review]

File: cachi2/core/scm.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (chmeliik) [reply cachi2/core/scm.py:37]: WDYT about the caching

For what it's worth, getting the repo id takes about 2.3 milliseconds on my machine, so if we run that once per package manager, or once per package, it's not the end of the world

```bash
$ python -m timeit --setup 'from cachi2.core.scm import get_repo_id' 'get_repo_id(".")'
100 loops, best of 5: 2.35 msec per loop
```

**Review by ejegrova — APPROVED** [review]

**Review by lkolacek — APPROVED** [review]
Looks great :+1:

**Pull request merged**